core/src/main/java/edu/wpi/grip/core/Pipeline.java

```diff
@@ -44,27 +57,108 @@
      * Remove everything in the pipeline
      */
     public void clear() {
-        this.steps.stream().collect(Collectors.toList()).forEach(this::removeStep);
-
+        new ArrayList<>(steps).forEach(this::removeStep);
+        // We collect the list first because the event modifies the list
         this.sources.stream()
                 .map(SourceRemovedEvent::new)
                 .collect(Collectors.toList())
                 .forEach(this.eventBus::post);
     }
 
+    private final <R> R readSourcesSafely(Function<List<Source>, R> sourceListFunction) {
+        return accessSafely(sourceLock.readLock(), Collections.unmodifiableList(sources), sourceListFunction);
+    }
+
+    /**
+     * Returns a snapshot of all of the sources in the pipeline.
+     *
+     * @return an Immutable copy of the sources at the current point in the pipeline.
+     * @see <a href="https://youtu.be/ZeO_J2OcHYM?t=16m35s">Why we use ImmutableList return type</a>
```

Comment:
Based on that talk, we should just use immutable collections for sources and steps.  When we modify the pipeline, we can replace the existing list with a modified copy, making all of this stuff simpler and have a more standard interface.

So, we should ask ourselves why we need something to be mutable instead of asking ourselves why it should be immutable.

Thoughts?